types/index.d.ts

```diff
@@ -2,11 +2,17 @@
 
 import {Transformer} from 'unified'
 import {Prefix, CreateElementLike} from 'hast-to-hyperscript'
+import {Node} from 'unist'
 
 declare namespace rehypeReact {
   type FragmentLike<T> = (props: any) => T | null
 
-  type ComponentLike<T> = (props: {[prop: string]: unknown}) => T | null
+  interface ComponentProps {
+    [prop: string]: unknown
+    node?: Node
```

Comment:
do we want to make this optional? :thinking: 
We know `node` will only be passed when `passNode` is `true`.

If we overload `rehypeReact` with two sets of options, one where `passNode` is true and one where `passNode` is omitted or false, we can clarify in the types whether `node` will or will not be passed as part of `ComponentProps`.

Reply:
It should work too. I'll change it.

Reply:
I've tried but I'm not sure it is worth to do. If users enable passNode, they need to update type of `props` argument of every `components` manually because our type inference for the argument is broken.


Before using passNode

```ts
unified().use(rehypeToReact, {
  createElement: React.createElement,
  components: {
    a: (props: rehypeToReact.ComponentProps) => <a>...</a>
  }
})
```

To access the `node` prop, we have to update argument type too(`ComponentProps` -> `ComponentPropsWithNode`) for every component using it.

```ts
unified().use(rehypeToReact, {
  createElement: React.createElement,
  passNode: true,
  components: {
  
    a: (props: rehypeToReact.ComponentPropsWithNode) => <a>{props.node}</a>
  }
})
```

Also we need to duplicate lots of interfaces like `ComponentPropsWithNode`, `ComponentLikeWithNo`, `OptionsWithPassNode` to add `node`.

Reply:
Anyawy, this is what I did try.

```ts
// Minimum TypeScript Version: 3.4

import {Transformer} from 'unified'
import {Prefix, CreateElementLike} from 'hast-to-hyperscript'
import {Node} from 'unist'

declare namespace rehypeReact {
  type FragmentLike<T> = (props: any) => T | null

  interface ComponentProps {
    [prop: string]: unknown
  }

  interface ComponentPropsWithNode {
    [prop: string]: unknown
    node: Node
  }

  type ComponentLike<T> = (props: ComponentProps) => T | null

  type ComponentLikeWithNode<T> = (props: ComponentPropsWithNode) => T | null

  interface Options<H extends CreateElementLike> {
    /**
     * How to create elements or components.
     * You should typically pass `React.createElement`
     */
    createElement: H

    /**
     * Create fragments instead of an outer `<div>` if available
     * You should typically pass `React.Fragment`
     */
    Fragment?: FragmentLike<ReturnType<H>>

    /**
     * Override default elements (such as `<a>`, `<p>`, etcetera) by passing an object mapping tag names to components
     */
    components?: {
      [element: string]: ComponentLike<ReturnType<H>>
    }

    /**
     * React key prefix
     *
     * @defaultValue 'h-'
     */
    prefix?: Prefix

    /**
     * Expose HAST Node objects to `node` prop of react components
     *
     * @defaultValue false
     */
    passNode?: false
  }

  interface OptionsWithPassNode<H extends CreateElementLike> {
    /**
     * How to create elements or components.
     * You should typically pass `React.createElement`
     */
    createElement: H

    /**
     * Create fragments instead of an outer `<div>` if available
     * You should typically pass `React.Fragment`
     */
    Fragment?: FragmentLike<ReturnType<H>>

    /**
     * Override default elements (such as `<a>`, `<p>`, etcetera) by passing an object mapping tag names to components
     */
    components?: {
      [element: string]: ComponentLikeWithNode<ReturnType<H>>
    }

    /**
     * React key prefix
     *
     * @defaultValue 'h-'
     */
    prefix?: Prefix

    /**
     * Expose HAST Node objects to `node` prop of react components
     *
     * @type boolean
     * @defaultValue false
     */
    passNode: true
  }
}

/**
 * Rehype plugin to transform to React
 */
declare function rehypeReact<H extends CreateElementLike>(
  options: rehypeReact.Options<H> | rehypeReact.OptionsWithPassNode<H>
): Transformer

export = rehypeReact
```

Reply:
> users enable passNode, they need to update type of props argument of every components manually because our type inference for the argument is broken

That is disappointing, let's skip this for now then.